pallet/staking/src/lib.rs

```diff
@@ -0,0 +1,918 @@
+
+	/// The map from (wannabe) collator to the preferences of that collator.
+	#[pallet::storage]
+	#[pallet::getter(fn collator_of)]
+	pub type Collators<T: Config> = StorageMap<_, Twox64Concat, T::AccountId, Perbill>;
+
+	/// Stakers' exposure.
+	#[pallet::storage]
+	#[pallet::unbounded]
+	#[pallet::getter(fn exposure)]
+	pub type Exposures<T: Config> =
+		StorageMap<_, Twox64Concat, T::AccountId, Exposure<T::AccountId>>;
+
+	/// The ideal number of active collators.
+	#[pallet::storage]
+	#[pallet::getter(fn collator_count)]
+	pub type CollatorCount<T> = StorageValue<_, u32, ValueQuery>;
+
+	/// The map from nominator to their nomination preferences, namely the collator that
+	/// they wish to support.
+	#[pallet::storage]
+	#[pallet::getter(fn nominator_of)]
+	pub type Nominators<T: Config> = StorageMap<_, Twox64Concat, T::AccountId, T::AccountId>;
+
+	/// Collator's reward points.
+	#[pallet::storage]
+	#[pallet::unbounded]
+	#[pallet::getter(fn reward_points)]
+	pub type RewardPoints<T: Config> =
+		StorageValue<_, (RewardPoint, BTreeMap<T::AccountId, RewardPoint>), ValueQuery>;
+
+	/// Active session's started timestamp.
+	#[pallet::storage]
+	#[pallet::getter(fn session_started_time)]
+	pub type SessionStartedTime<T: Config> = StorageValue<_, Moment, ValueQuery>;
+
+	/// Elapsed time.
+	#[pallet::storage]
+	#[pallet::getter(fn elapsed_time)]
+	pub type ElapsedTime<T: Config> = StorageValue<_, Moment, ValueQuery>;
+
+	#[pallet::genesis_config]
+	pub struct GenesisConfig<T: Config> {
+		/// Current timestamp.
+		pub now: Moment,
+		/// The running time of Darwinia1.
+		pub elapsed_time: Moment,
+		/// Genesis collator count.
+		pub collator_count: u32,
+		/// Genesis collator preferences.
+		pub collators: Vec<(T::AccountId, Balance)>,
+	}
+	#[cfg(feature = "std")]
+	impl<T: Config> Default for GenesisConfig<T> {
+		fn default() -> Self {
+			GenesisConfig { now: 0, elapsed_time: 0, collator_count: 0, collators: Vec::new() }
+		}
+	}
+	#[pallet::genesis_build]
+	impl<T: Config> GenesisBuild<T> for GenesisConfig<T> {
+		fn build(&self) {
+			<SessionStartedTime<T>>::put(self.now);
+			<ElapsedTime<T>>::put(self.elapsed_time);
+			<CollatorCount<T>>::put(self.collator_count);
+
+			self.collators.iter().cloned().for_each(|(who, stake)| {
+				<Pallet<T>>::stake(RawOrigin::Signed(who.clone()).into(), stake, 0, Vec::new())
+					.unwrap();
+				<Pallet<T>>::collect(RawOrigin::Signed(who).into(), Default::default()).unwrap();
+			});
+		}
+	}
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	pub struct Pallet<T>(PhantomData<T>);
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Add stakes to the staking pool.
+		///
+		/// This will transfer the stakes to a pallet/contact account.
+		#[pallet::weight(0)]
+		pub fn stake(
+			origin: OriginFor<T>,
+			ring_amount: Balance,
+			kton_amount: Balance,
+			deposits: Vec<DepositId<T>>,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
```

Comment:
Note that this can also be called using substrate dispatch call, the signature is `pub struct EthereumSignature(ecdsa::Signature);`

Normally, it is designed to be called by precompiles.

Reply:
We could keep the dispatch call because of the following: 
1. Easy to debug
2. Easy to benchmark
3. Transactional call